pkg/storage/replica.go

```diff
@@ -1143,15 +1144,104 @@ func (r *Replica) checkBatchRequest(ba roachpb.BatchRequest) error {
-			spans[i] = union.GetInner().Header()
+			args := union.GetInner()
+			spans[i] = args.Header()
+			switch t := args.(type) {
+			case *roachpb.PushTxnRequest:
+				spans = append(spans, roachpb.Span{
+					Key: keys.AbortCacheKey(r.RangeID, t.PusheeTxn.ID),
+				}, roachpb.Span{
+					Key: keys.TransactionKey(t.PusheeTxn.Key, t.PusheeTxn.ID),
+				})
+			case *roachpb.ResolveIntentRequest:
+				spans = append(spans, roachpb.Span{
+					Key: keys.AbortCacheKey(r.RangeID, t.IntentTxn.ID),
+				})
+			case *roachpb.ResolveIntentRangeRequest:
+				spans = append(spans, roachpb.Span{
+					Key: keys.AbortCacheKey(r.RangeID, t.IntentTxn.ID),
+				})
+			case *roachpb.GCRequest:
+				for _, key := range t.Keys {
+					fmt.Printf("gc: %v", key)
+					spans = append(spans, roachpb.Span{
+						Key:    key.Key,
+						EndKey: key.Key.Next(),
+					})
+				}
+				spans = append(spans, roachpb.Span{
+					Key: keys.RangeLastGCKey(r.RangeID),
+				}, roachpb.Span{
+					Key: keys.RangeLastGCKey(r.RangeID),
+				}, roachpb.Span{
+					Key: keys.RangeTxnSpanGCThresholdKey(r.RangeID),
+				})
+			case *roachpb.EndTransactionRequest:
+				// The spans may extend beyond this Range, but it's ok for the
+				// purpose of the command queue.
+				spans = append(spans, t.IntentSpans...)
+				trigger := t.InternalCommitTrigger
+				if trigger == nil {
+					break
+				}
+				if trigger.SplitTrigger != nil {
+					spans = append(spans, roachpb.Span{
+						Key:    keys.MakeRangeIDPrefix(trigger.SplitTrigger.RightDesc.RangeID),
+						EndKey: keys.MakeRangeIDPrefix(trigger.SplitTrigger.RightDesc.RangeID + 1),
+					}, roachpb.Span{
+						// Need everything here since we recompute everything
+						// sometimes.
+						Key:    trigger.SplitTrigger.LeftDesc.StartKey.AsRawKey(),
+						EndKey: trigger.SplitTrigger.RightDesc.EndKey.AsRawKey(),
+					}, roachpb.Span{
+						Key:    r.abortCache.min(),
+						EndKey: r.abortCache.max(),
+					})
+				} else if trigger.MergeTrigger != nil {
+					spans = append(spans, roachpb.Span{
+						Key:    keys.MakeRangeIDPrefix(trigger.MergeTrigger.RightDesc.RangeID),
+						EndKey: keys.MakeRangeIDPrefix(trigger.MergeTrigger.RightDesc.RangeID + 1),
+					}, roachpb.Span{
+						Key:    r.abortCache.min(),
+						EndKey: r.abortCache.max(),
+					}, roachpb.Span{
+						// Interesting that we don't see this one pop up always;
+						// this key can certainly not be added to the cmdq in
+						// general as that would make it block everything on
+						// everything else (except reads on reads).
+						Key: keys.RangeStatsKey(r.RangeID),
+					}, roachpb.Span{
+						Key:    trigger.MergeTrigger.RightDesc.StartKey.AsRawKey(),
+						EndKey: trigger.MergeTrigger.RightDesc.EndKey.AsRawKey(),
+					})
+				}
+			}
 		}
+
+		if ba.Txn != nil {
+			spans = append(spans, roachpb.Span{
+				Key: keys.AbortCacheKey(r.RangeID, ba.Txn.ID),
+			}, roachpb.Span{
+				Key: keys.TransactionKey(ba.Txn.Key, ba.Txn.ID),
```

Comment:
I understand now why we need to read the abort cache key, but we shouldn't be reading the transaction record on every command - it might not be present on this range.